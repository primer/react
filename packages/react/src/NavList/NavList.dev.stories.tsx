import type {Meta} from '@storybook/react'
import React from 'react'
import {PageLayout} from '../PageLayout'
import {NavList} from './NavList'
import {ArrowRightIcon, ArrowLeftIcon, BookIcon, FileDirectoryIcon} from '@primer/octicons-react'

const meta: Meta = {
  title: 'Components/NavList/DevOnly',
  component: NavList,
  parameters: {
    layout: 'fullscreen',
  },
}

export const WithBadExampleOfSubNavAndTrailingAction = () => {
  return (
    <PageLayout>
      <PageLayout.Pane position="start">
        <NavList>
          <NavList.Item>
            <NavList.LeadingVisual>
              <FileDirectoryIcon />
            </NavList.LeadingVisual>
            Item 1
            <NavList.TrailingAction label="Expand sidebar" icon={ArrowLeftIcon} />
          </NavList.Item>
          <NavList.Item>
            Item 2
            <NavList.TrailingAction as="a" href="#" label="Some action" icon={ArrowRightIcon} />
          </NavList.Item>
          <NavList.Item>
            Item 3
            <NavList.TrailingAction label="This is not supported and should not render!!!!!" icon={BookIcon} />
            <NavList.SubNav>
              <NavList.Item href="#">
                Sub item 1
                <NavList.TrailingAction label="Another action" icon={BookIcon} />
              </NavList.Item>
            </NavList.SubNav>
          </NavList.Item>
        </NavList>
      </PageLayout.Pane>
    </PageLayout>
  )
}

WithBadExampleOfSubNavAndTrailingAction.storyName = 'With SubNav and Trailing Action (Bad example, do not copy)'

export default meta
