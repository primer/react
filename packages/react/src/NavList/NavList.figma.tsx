import React from 'react'
import {NavList} from '../../src'
import figma from '@figma/code-connect'

figma.connect(
  NavList,
  'https://www.figma.com/design/GCvY3Qv8czRgZgvl1dG6lp/Primer-Web?node-id=16716-62541&t=m8uYul4RVKTAkjzl-4',
  {
    props: {
      children: figma.children('*'),
    },
    example: ({children}) => <NavList>{children}</NavList>,
  },
)
