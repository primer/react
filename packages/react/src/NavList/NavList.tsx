import {ChevronDownIcon} from '@primer/octicons-react'
import type {ForwardRefComponent as PolymorphicForwardRefComponent} from '../utils/polymorphic'
import React, {isValidElement} from 'react'
import styled from 'styled-components'
import type {ActionListDividerProps, ActionListLeadingVisualProps, ActionListTrailingVisualProps} from '../ActionList'
import {ActionList} from '../ActionList'
import {ActionListContainerContext} from '../ActionList/ActionListContainerContext'
import Box from '../Box'
import Octicon from '../Octicon'
import type {SxProp} from '../sx'
import sx, {merge} from '../sx'
import {defaultSxProp} from '../utils/defaultSxProp'
import {useId} from '../hooks/useId'
import useIsomorphicLayoutEffect from '../utils/useIsomorphicLayoutEffect'

const getSubnavStyles = (depth: number) => {
  return {
    paddingLeft: depth > 0 ? depth + 2 : null, // Indent sub-items
    fontSize: depth > 0 ? 0 : null, // Reduce font size of sub-items
    fontWeight: depth > 0 ? 'normal' : null, // Sub-items don't get bolded
  }
}

// ----------------------------------------------------------------------------
// NavList

export type NavListProps = {
  children: React.ReactNode
} & SxProp &
  React.ComponentProps<'nav'>

const NavBox = styled.nav<SxProp>(sx)

const Root = React.forwardRef<HTMLElement, NavListProps>(({children, ...props}, ref) => {
  return (
    <NavBox {...props} ref={ref}>
      <ActionListContainerContext.Provider
        value={{
          container: 'NavList',
        }}
      >
        <ActionList>{children}</ActionList>
      </ActionListContainerContext.Provider>
    </NavBox>
  )
})

Root.displayName = 'NavList'

// ----------------------------------------------------------------------------
// NavList.Item

export type NavListItemProps = {
  children: React.ReactNode
  defaultOpen?: boolean
  href?: string
  'aria-current'?: 'page' | 'step' | 'location' | 'date' | 'time' | 'true' | 'false' | boolean
  inactiveText?: string
} & SxProp

const Item = React.forwardRef<HTMLAnchorElement, NavListItemProps>(
  ({'aria-current': ariaCurrent, children, defaultOpen, sx: sxProp = defaultSxProp, ...props}, ref) => {
    const {depth} = React.useContext(SubNavContext)

    // Get SubNav from children
    const subNav = React.Children.toArray(children).find(child => isValidElement(child) && child.type === SubNav)

    // Get children without SubNav
    const childrenWithoutSubNav = React.Children.toArray(children).filter(child =>
      isValidElement(child) ? child.type !== SubNav : true,
    )

    if (!isValidElement(subNav) && defaultOpen)
      // eslint-disable-next-line no-console
      console.error('NavList.Item must have a NavList.SubNav to use defaultOpen.')

    // Render ItemWithSubNav if SubNav is present
    if (subNav && isValidElement(subNav)) {
      return (
        <ItemWithSubNav subNav={subNav} depth={depth} defaultOpen={defaultOpen} sx={sxProp}>
          {childrenWithoutSubNav}
        </ItemWithSubNav>
      )
    }

    return (
      <ActionList.LinkItem
        ref={ref}
        aria-current={ariaCurrent}
        active={Boolean(ariaCurrent) && ariaCurrent !== 'false'}
        sx={merge<SxProp['sx']>(getSubnavStyles(depth), sxProp)}
        {...props}
      >
        {children}
      </ActionList.LinkItem>
    )
  },
) as PolymorphicForwardRefComponent<'a', NavListItemProps>

Item.displayName = 'NavList.Item'

// ----------------------------------------------------------------------------
// ItemWithSubNav (internal)

type ItemWithSubNavProps = {
  children: React.ReactNode
  subNav: React.ReactNode
  depth: number
  defaultOpen?: boolean
} & SxProp

const ItemWithSubNavContext = React.createContext<{buttonId: string; subNavId: string; isOpen: boolean}>({
  buttonId: '',
  subNavId: '',
  isOpen: false,
})

// TODO: ref prop
// TODO: Animate open/close transition
function ItemWithSubNav({children, subNav, depth, defaultOpen, sx: sxProp = defaultSxProp}: ItemWithSubNavProps) {
  const buttonId = useId()
  const subNavId = useId()
  const [isOpen, setIsOpen] = React.useState((defaultOpen || null) ?? false)
  const subNavRef = React.useRef<HTMLDivElement>(null)
  const [containsCurrentItem, setContainsCurrentItem] = React.useState(false)

  useIsomorphicLayoutEffect(() => {
    if (subNavRef.current) {
      // Check if SubNav contains current item
      // valid values: page, step, location, date, time, true and false
      const currentItem = subNavRef.current.querySelector('[aria-current]:not([aria-current=false])')

      if (currentItem) {
        setContainsCurrentItem(true)
        setIsOpen(true)
      }
    }
  }, [subNav, buttonId])

  return (
    <ItemWithSubNavContext.Provider value={{buttonId, subNavId, isOpen}}>
      <Box as="li" aria-labelledby={buttonId} sx={{listStyle: 'none'}}>
        <ActionList.Item
          as="button"
          id={buttonId}
          aria-expanded={isOpen}
          aria-controls={subNavId}
          // When the subNav is closed, how should we indicated that the subNav contains the current item?
          active={!isOpen && containsCurrentItem}
          onClick={() => setIsOpen(open => !open)}
          sx={merge<SxProp['sx']>(
            {
              ...getSubnavStyles(depth),
              fontWeight: containsCurrentItem ? 'bold' : null, // Parent item is bold if any of it's sub-items are current
            },
            sxProp,
          )}
        >
          {children}
          {/* What happens if the user provides a TrailingVisual? */}
          <ActionList.TrailingVisual>
            <Octicon
              icon={ChevronDownIcon}
              sx={{
                transform: isOpen ? 'rotate(180deg)' : 'rotate(0deg)',
              }}
            />
          </ActionList.TrailingVisual>
        </ActionList.Item>

        <div ref={subNavRef}>{subNav}</div>
      </Box>
    </ItemWithSubNavContext.Provider>
  )
}

// ----------------------------------------------------------------------------
// NavList.SubNav

export type NavListSubNavProps = {
  children: React.ReactNode
} & SxProp

const SubNavContext = React.createContext<{depth: number}>({depth: 0})

// TODO: ref prop
// NOTE: SubNav must be a direct child of an Item
const SubNav = ({children, sx: sxProp = defaultSxProp}: NavListSubNavProps) => {
  const {buttonId, subNavId, isOpen} = React.useContext(ItemWithSubNavContext)
  const {depth} = React.useContext(SubNavContext)

  if (!buttonId || !subNavId) {
    // eslint-disable-next-line no-console
    console.error('NavList.SubNav must be a child of a NavList.Item')
  }

  if (depth > 3) {
    // TODO: write a more informative error message that directs people to rethink their IA
    // eslint-disable-next-line no-console
    console.error('NavList.SubNav only supports four levels of nesting')
    return null
  }

  return (
    <SubNavContext.Provider value={{depth: depth + 1}}>
      <Box
        as="ul"
        id={subNavId}
        aria-labelledby={buttonId}
        sx={merge<SxProp['sx']>(
          {
            padding: 0,
            margin: 0,
            display: isOpen ? 'block' : 'none',
          },
          sxProp,
        )}
      >
        {children}
      </Box>
    </SubNavContext.Provider>
  )
}

SubNav.displayName = 'NavList.SubNav'

// ----------------------------------------------------------------------------
// NavList.LeadingVisual

export type NavListLeadingVisualProps = ActionListLeadingVisualProps

const LeadingVisual = ActionList.LeadingVisual

LeadingVisual.displayName = 'NavList.LeadingVisual'

// ----------------------------------------------------------------------------
// NavList.TrailingVisual

export type NavListTrailingVisualProps = ActionListTrailingVisualProps

const TrailingVisual = ActionList.TrailingVisual

TrailingVisual.displayName = 'NavList.TrailingVisual'

// ----------------------------------------------------------------------------
// NavList.Divider

export type NavListDividerProps = ActionListDividerProps

const Divider = ActionList.Divider

Divider.displayName = 'NavList.Divider'

// ----------------------------------------------------------------------------
// NavList.Group

export type NavListGroupProps = {
  children: React.ReactNode
  title?: string
} & SxProp

const defaultSx = {}
// TODO: ref prop
const Group: React.FC<NavListGroupProps> = ({title, children, sx: sxProp = defaultSx, ...props}) => {
  return (
    <>
      {/* Hide divider if the group is the first item in the list */}
      <ActionList.Divider sx={{'&:first-child': {display: 'none'}}} />
      <ActionList.Group {...props} sx={sxProp}>
        {/* Setting up the default value for the heading level. TODO: API update to give flexibility to NavList.Group title's heading level */}
        {title ? <ActionList.GroupHeading as="h3">{title}</ActionList.GroupHeading> : null}
        {children}
      </ActionList.Group>
    </>
  )
}

Group.displayName = 'NavList.Group'

// ----------------------------------------------------------------------------
// Export

export const NavList = Object.assign(Root, {
  Item,
  SubNav,
  LeadingVisual,
  TrailingVisual,
  Divider,
  Group,
})
