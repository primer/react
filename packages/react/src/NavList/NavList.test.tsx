import {render, fireEvent} from '@testing-library/react'
import React from 'react'
import {ThemeProvider} from '..'
import {NavList} from './NavList'
import {FeatureFlags} from '../FeatureFlags'

type ReactRouterLikeLinkProps = {to: string; children: React.ReactNode}

const ReactRouterLikeLink = React.forwardRef<HTMLAnchorElement, ReactRouterLikeLinkProps>(
  ({to, children, ...props}, ref) => {
    return (
      <a ref={ref} href={to} {...props}>
        {children}
      </a>
    )
  },
)

type NextJSLinkProps = {href: string; children: React.ReactNode}

const NextJSLikeLink = React.forwardRef<HTMLAnchorElement, NextJSLinkProps>(
  ({href, children}, ref): React.ReactElement => {
    const child = React.Children.only(children)
    const childProps = {
      ref,
      href,
    }
    return <>{React.isValidElement(child) ? React.cloneElement(child, childProps) : null}</>
  },
)

describe('NavList', () => {
  it('renders a simple list', () => {
    const {container} = render(
      <ThemeProvider>
        <NavList>
          <NavList.Item href="/" aria-current="page">
            Home
          </NavList.Item>
          <NavList.Item href="/about">About</NavList.Item>
          <NavList.Item href="/contact">Contact</NavList.Item>
        </NavList>
      </ThemeProvider>,
    )
    expect(container).toMatchSnapshot()
  })

  it('renders with groups', () => {
    const {container} = render(
      <ThemeProvider>
        <NavList>
          <NavList.Group title="Overview">
            <NavList.Item href="/getting-started" aria-current="page">
              Getting started
            </NavList.Item>
          </NavList.Group>
          <NavList.Group title="Components">
            <NavList.Item href="/Avatar">Avatar</NavList.Item>
          </NavList.Group>
        </NavList>
      </ThemeProvider>,
    )
    expect(container).toMatchSnapshot()
  })

  it('supports TrailingAction', async () => {
    const {getByRole} = render(
      <NavList>
        <NavList.Item>
          Item 1
          <NavList.TrailingAction label="Some trailing action" />
        </NavList.Item>
      </NavList>,
    )

    const trailingAction = getByRole('button', {name: 'Some trailing action'})
    expect(trailingAction).toBeInTheDocument()
  })
})

describe('NavList.Item', () => {
  it('passes aria-current prop to the underlying link', () => {
    const {getByRole} = render(
      <NavList>
        <NavList.Item href="/" aria-current="page">
          Home
        </NavList.Item>
        <NavList.Item href="/about">About</NavList.Item>
        <NavList.Item href="/contact">Contact</NavList.Item>
      </NavList>,
    )

    const homeLink = getByRole('link', {name: 'Home'})
    const aboutLink = getByRole('link', {name: 'About'})

    expect(homeLink).toHaveAttribute('aria-current', 'page')
    expect(aboutLink).not.toHaveAttribute('aria-current')
  })

  it('is compatiable with React-Router-like link components', () => {
    const {getByRole} = render(
      <NavList>
        <NavList.Item as={ReactRouterLikeLink} to={'/'} aria-current="page">
          React Router link
        </NavList.Item>
      </NavList>,
    )

    const link = getByRole('link', {name: 'React Router link'})

    expect(link).toHaveAttribute('aria-current', 'page')
    expect(link).toHaveAttribute('href', '/')
  })

  it('is compatible with NextJS-like link components', () => {
    const {getByRole} = render(
      <NavList>
        <NextJSLikeLink href="/">
          <NavList.Item aria-current="page">NextJS link</NavList.Item>
        </NextJSLikeLink>
      </NavList>,
    )

    const link = getByRole('link', {name: 'NextJS link'})

    expect(link).toHaveAttribute('href', '/')
    expect(link).toHaveAttribute('aria-current', 'page')
  })
})

describe('NavList.Item with NavList.SubNav', () => {
  function NavListWithSubNav() {
    return (
      <NavList>
        <NavList.Item href="#">Item 1</NavList.Item>
        <NavList.Item href="#">
          Item 2
          <NavList.SubNav>
            <NavList.Item href="#">Sub Item 1</NavList.Item>
            <NavList.Item href="#">Sub Item 2</NavList.Item>
          </NavList.SubNav>
        </NavList.Item>
        <NavList.Item href="#">Item 3</NavList.Item>
      </NavList>
    )
  }

  function NavListWithCurrentSubNav() {
    return (
      <NavList>
        <NavList.Item href="#">Item 1</NavList.Item>
        <NavList.Item href="#">
          Item 2
          <NavList.SubNav>
            <NavList.Item href="#" aria-current="page">
              Sub Item 1
            </NavList.Item>
            <NavList.Item href="#">Sub Item 2</NavList.Item>
          </NavList.SubNav>
        </NavList.Item>
        <NavList.Item href="#">Item 3</NavList.Item>
      </NavList>
    )
  }

  it('renders as a button', () => {
    const {queryByRole} = render(<NavListWithSubNav />)
    const itemWithSubNav = queryByRole('button', {name: 'Item 2'})
    expect(itemWithSubNav).toBeInTheDocument()
  })

  it('ignores aria-current prop', () => {
    const {queryByRole} = render(<NavListWithSubNav />)
    const itemWithSubNav = queryByRole('button', {name: 'Item 2'})
    expect(itemWithSubNav).not.toHaveAttribute('aria-current')
  })

  it('shows SubNav by default if SubNav contains the current item', () => {
    const {queryByRole, getByRole} = render(<NavListWithCurrentSubNav />)

    // Starts open
    expect(queryByRole('list', {name: 'Item 2'})).toBeVisible()

    // Click to close
    const itemWithSubNav = getByRole('button', {name: 'Item 2'})
    fireEvent.click(itemWithSubNav)
    expect(queryByRole('list', {name: 'Item 2'})).toBeNull()
  })

  it('hides SubNav by default if SubNav does not contain the current item', () => {
    const {queryByRole} = render(<NavListWithSubNav />)
    const subNav = queryByRole('list', {name: 'Item 2'})
    expect(subNav).toBeNull()
  })

  it('toggles visiblility of SubNav when clicked', () => {
    const {getByRole, queryByRole} = render(<NavListWithSubNav />)
    const itemWithSubNav = getByRole('button', {name: 'Item 2'})

    // Starts closed
    expect(queryByRole('list', {name: 'Item 2'})).toBeNull()

    // Click to open
    fireEvent.click(itemWithSubNav)
    expect(queryByRole('list', {name: 'Item 2'})).toBeVisible()

    // Click to close
    fireEvent.click(itemWithSubNav)
    expect(queryByRole('list', {name: 'Item 2'})).toBeNull()
  })

  it('has active styles if SubNav contains the current item and is closed', () => {
    const {container, getByRole, queryByRole} = render(
      <ThemeProvider>
        <NavList>
          <NavList.Item>
            Item
            <NavList.SubNav>
              <NavList.Item href="#" aria-current="page">
                Sub Item
              </NavList.Item>
            </NavList.SubNav>
          </NavList.Item>
        </NavList>
      </ThemeProvider>,
    )

    const button = getByRole('button')

    // Starts open
    expect(queryByRole('list', {name: 'Item'})).toBeVisible()

    // Click to close
    fireEvent.click(button)
    expect(queryByRole('list', {name: 'Item'})).toBeNull()

    // Snapshot styles
    expect(container).toMatchSnapshot()
  })

  it('does not have active styles if SubNav contains the current item and is open', () => {
    const {container, queryByRole} = render(
      <ThemeProvider>
        <NavList>
          <NavList.Item>
            Item
            <NavList.SubNav>
              <NavList.Item href="#" aria-current="page">
                Sub Item
              </NavList.Item>
            </NavList.SubNav>
          </NavList.Item>
        </NavList>
      </ThemeProvider>,
    )

    // Starts open
    expect(queryByRole('list', {name: 'Item'})).toBeVisible()

    // Snapshot styles
    expect(container).toMatchSnapshot()
  })

  it('prevents more than 4 levels of nested SubNavs', () => {
    const consoleSpy = jest
      .spyOn(console, 'error')
      // Suppress error message in test output
      .mockImplementation(() => null)

    const {getByRole} = render(
      <NavList>
        <NavList.Item href="#">Item 1</NavList.Item>
        <NavList.Item href="#">
          Item 2{/* NOTE: Don't nest SubNavs. For testing purposes only */}
          <NavList.SubNav>
            <NavList.Item href="#">
              Sub item 1
              <NavList.SubNav>
                <NavList.Item href="#">
                  Sub item 1.1
                  <NavList.SubNav>
                    <NavList.Item href="#">Sub item 1.1.1</NavList.Item>
                    <NavList.Item href="#">
                      Sub item 1.1.2
                      <NavList.SubNav>
                        <NavList.Item href="#">Sub item 1.1.2.1</NavList.Item>
                        <NavList.Item href="#">
                          Sub item 1.1.2.2
                          <NavList.SubNav>
                            <NavList.Item href="#" aria-current="page">
                              Sub item 1.1.2.2.1
                            </NavList.Item>
                            <NavList.Item href="#">Sub item 1.1.2.2.2</NavList.Item>
                          </NavList.SubNav>
                        </NavList.Item>
                      </NavList.SubNav>
                    </NavList.Item>
                  </NavList.SubNav>
                </NavList.Item>
                <NavList.Item href="#">Sub item 1.2</NavList.Item>
              </NavList.SubNav>
            </NavList.Item>
            <NavList.Item href="#">Sub item 2</NavList.Item>
          </NavList.SubNav>
        </NavList.Item>
        <NavList.Item href="#">Item 3</NavList.Item>
      </NavList>,
    )

    const item = getByRole('button', {name: 'Item 2'})
    fireEvent.click(item)

    expect(consoleSpy).toHaveBeenCalled()
  })

  it('is compatiable with React-Router-like link components', () => {
    function NavLink({href, children}: {href: string; children: React.ReactNode}) {
      // In a real app, you'd check if the href matches the url of the current page. For testing purposes, we'll use the text of the link to determine if it's current
      const isCurrent = children === 'Current'
      return (
        <NavList.Item as={ReactRouterLikeLink} to={href} aria-current={isCurrent ? 'page' : false}>
          {children}
        </NavList.Item>
      )
    }

    const {queryByRole} = render(
      <NavList>
        <NavLink href="/">Item 1</NavLink>
        <NavList.Item>
          Item 2
          <NavList.SubNav>
            <NavLink href="/sub-item-1">Current</NavLink>
            <NavLink href="/sub-item-2">Sub item 2</NavLink>
          </NavList.SubNav>
        </NavList.Item>
        <NavLink href="/">Item 3</NavLink>
      </NavList>,
    )

    const currentLink = queryByRole('link', {name: 'Current'})
    expect(currentLink).toBeVisible()
  })

  describe('TrailingAction', () => {
    function NavListWithSubNavAndTrailingAction() {
      return (
        <FeatureFlags flags={{primer_react_action_list_item_as_button: true}}>
          <NavList>
            <NavList.Item href="#">
              Item
              <NavList.TrailingAction label="This should not be rendered" />
              <NavList.SubNav>
                <NavList.Item href="#">
                  Sub Item 1
                  <NavList.TrailingAction label="Trailing Action for Sub Item 1" />
                </NavList.Item>
                <NavList.Item href="#">Sub Item 2</NavList.Item>
              </NavList.SubNav>
            </NavList.Item>
          </NavList>
        </FeatureFlags>
      )
    }

    it('does not render TrailingAction alongside SubNav', async () => {
      const {queryByRole} = render(<NavListWithSubNavAndTrailingAction />)

      const trailingAction = queryByRole('button', {name: 'This should not be rendered'})
      expect(trailingAction).toBeNull()
    })

    it('supports TrailingAction within an Item inside SubNav', async () => {
      const {getByRole, queryByRole} = render(<NavListWithSubNavAndTrailingAction />)

      const itemWithSubNav = getByRole('button', {name: 'Item'})
      fireEvent.click(itemWithSubNav)

      expect(queryByRole('link', {name: 'Sub Item 1'})).toBeVisible()
      expect(queryByRole('button', {name: 'Trailing Action for Sub Item 1'})).toBeVisible()
    })
  })
})
