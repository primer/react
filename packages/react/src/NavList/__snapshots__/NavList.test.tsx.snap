// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavList > renders a simple list 1`] = `
<div>
  <nav>
    <ul
      class="prc-ActionList-ActionList--dkHj"
      data-dividers="false"
      data-variant="inset"
    >
      <li
        class="prc-ActionList-ActionListItem-jXerx"
        data-active="true"
        data-has-description="false"
      >
        <a
          aria-current="page"
          aria-labelledby=":r1:--label  "
          class="prc-ActionList-ActionListContent-Q8eQr prc-Link-Link-O33ci"
          data-size="medium"
          href="/"
          id=":r1:"
          style="--subitem-depth: 0;"
          tabindex="0"
        >
          <span
            class="prc-ActionList-Spacer-vgXtU"
          />
          <span
            class="prc-ActionList-ActionListSubContent-Zcjdz"
            data-component="ActionList.Item--DividerContainer"
          >
            <span
              class="prc-ActionList-ItemLabel-hYyH7"
              id=":r1:--label"
            >
              Home
            </span>
          </span>
        </a>
      </li>
      <li
        class="prc-ActionList-ActionListItem-jXerx"
        data-has-description="false"
      >
        <a
          aria-labelledby=":r2:--label  "
          class="prc-ActionList-ActionListContent-Q8eQr prc-Link-Link-O33ci"
          data-size="medium"
          href="/about"
          id=":r2:"
          style="--subitem-depth: 0;"
          tabindex="0"
        >
          <span
            class="prc-ActionList-Spacer-vgXtU"
          />
          <span
            class="prc-ActionList-ActionListSubContent-Zcjdz"
            data-component="ActionList.Item--DividerContainer"
          >
            <span
              class="prc-ActionList-ItemLabel-hYyH7"
              id=":r2:--label"
            >
              About
            </span>
          </span>
        </a>
      </li>
      <li
        class="prc-ActionList-ActionListItem-jXerx"
        data-has-description="false"
      >
        <a
          aria-labelledby=":r3:--label  "
          class="prc-ActionList-ActionListContent-Q8eQr prc-Link-Link-O33ci"
          data-size="medium"
          href="/contact"
          id=":r3:"
          style="--subitem-depth: 0;"
          tabindex="0"
        >
          <span
            class="prc-ActionList-Spacer-vgXtU"
          />
          <span
            class="prc-ActionList-ActionListSubContent-Zcjdz"
            data-component="ActionList.Item--DividerContainer"
          >
            <span
              class="prc-ActionList-ItemLabel-hYyH7"
              id=":r3:--label"
            >
              Contact
            </span>
          </span>
        </a>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`NavList > renders with groups 1`] = `
<div>
  <nav>
    <ul
      class="prc-ActionList-ActionList--dkHj"
      data-dividers="false"
      data-variant="inset"
    >
      <li
        aria-hidden="true"
        class="prc-ActionList-Divider-eZxO8"
        data-component="ActionList.Divider"
      />
      <li
        class="prc-ActionList-Group-42hnW"
      >
        <div
          class="prc-ActionList-GroupHeadingWrap-CL7FP"
          data-component="GroupHeadingWrap"
          data-variant="subtle"
        >
          <h3
            class="prc-ActionList-GroupHeading-w2osB"
            data-component="ActionList.GroupHeading"
            id=":r5:"
          >
            Overview
          </h3>
        </div>
        <ul
          aria-labelledby=":r5:"
          class="prc-ActionList-GroupList-kpPZc"
        >
          <li
            class="prc-ActionList-ActionListItem-jXerx"
            data-active="true"
            data-has-description="false"
          >
            <a
              aria-current="page"
              aria-labelledby=":r6:--label  "
              class="prc-ActionList-ActionListContent-Q8eQr prc-Link-Link-O33ci"
              data-size="medium"
              href="/getting-started"
              id=":r6:"
              style="--subitem-depth: 0;"
              tabindex="0"
            >
              <span
                class="prc-ActionList-Spacer-vgXtU"
              />
              <span
                class="prc-ActionList-ActionListSubContent-Zcjdz"
                data-component="ActionList.Item--DividerContainer"
              >
                <span
                  class="prc-ActionList-ItemLabel-hYyH7"
                  id=":r6:--label"
                >
                  Getting started
                </span>
              </span>
            </a>
          </li>
        </ul>
      </li>
      <li
        aria-hidden="true"
        class="prc-ActionList-Divider-eZxO8"
        data-component="ActionList.Divider"
      />
      <li
        class="prc-ActionList-Group-42hnW"
      >
        <div
          class="prc-ActionList-GroupHeadingWrap-CL7FP"
          data-component="GroupHeadingWrap"
          data-variant="subtle"
        >
          <h3
            class="prc-ActionList-GroupHeading-w2osB"
            data-component="ActionList.GroupHeading"
            id=":r7:"
          >
            Components
          </h3>
        </div>
        <ul
          aria-labelledby=":r7:"
          class="prc-ActionList-GroupList-kpPZc"
        >
          <li
            class="prc-ActionList-ActionListItem-jXerx"
            data-has-description="false"
          >
            <a
              aria-labelledby=":r8:--label  "
              class="prc-ActionList-ActionListContent-Q8eQr prc-Link-Link-O33ci"
              data-size="medium"
              href="/Avatar"
              id=":r8:"
              style="--subitem-depth: 0;"
              tabindex="0"
            >
              <span
                class="prc-ActionList-Spacer-vgXtU"
              />
              <span
                class="prc-ActionList-ActionListSubContent-Zcjdz"
                data-component="ActionList.Item--DividerContainer"
              >
                <span
                  class="prc-ActionList-ItemLabel-hYyH7"
                  id=":r8:--label"
                >
                  Avatar
                </span>
              </span>
            </a>
          </li>
        </ul>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`NavList.Item with NavList.SubNav > does not have active styles if SubNav contains the current item and is open 1`] = `
<div>
  <nav>
    <ul
      class="prc-ActionList-ActionList--dkHj"
      data-dividers="false"
      data-variant="inset"
    >
      <li
        class="prc-ActionList-ActionListItem-jXerx"
        data-has-description="false"
        data-has-subitem="true"
      >
        <button
          aria-controls=":r23:"
          aria-expanded="true"
          aria-labelledby=":r22:--label :r22:--trailing-visual "
          class="prc-ActionList-ActionListContent-Q8eQr"
          data-size="medium"
          id=":r22:"
          style="--subitem-depth: 0;"
          tabindex="0"
          type="button"
        >
          <span
            class="prc-ActionList-Spacer-vgXtU"
          />
          <span
            class="prc-ActionList-ActionListSubContent-Zcjdz"
            data-component="ActionList.Item--DividerContainer"
          >
            <span
              class="prc-ActionList-ItemLabel-hYyH7"
              id=":r22:--label"
            >
              Item
            </span>
            <span
              class="prc-ActionList-TrailingVisual-8LMjP prc-ActionList-VisualWrap-MGVnU"
              id=":r22:--trailing-visual"
            >
              <svg
                aria-hidden="true"
                class="octicon octicon-chevron-down prc-ActionList-ExpandIcon-6JuzS"
                display="inline-block"
                fill="currentColor"
                focusable="false"
                height="16"
                overflow="visible"
                style="vertical-align: text-bottom;"
                viewBox="0 0 16 16"
                width="16"
              >
                <path
                  d="M12.78 5.22a.749.749 0 0 1 0 1.06l-4.25 4.25a.749.749 0 0 1-1.06 0L3.22 6.28a.749.749 0 1 1 1.06-1.06L8 8.939l3.72-3.719a.749.749 0 0 1 1.06 0Z"
                />
              </svg>
            </span>
          </span>
        </button>
        <ul
          aria-labelledby=":r22:"
          class="prc-ActionList-SubGroup-4pVPh"
          id=":r23:"
        >
          <li
            class="prc-ActionList-ActionListItem-jXerx"
            data-active="true"
            data-has-description="false"
          >
            <a
              aria-current="page"
              aria-labelledby=":r25:--label  "
              class="prc-ActionList-ActionListContent-Q8eQr prc-Link-Link-O33ci"
              data-size="medium"
              href="#"
              id=":r25:"
              style="--subitem-depth: 1;"
              tabindex="0"
            >
              <span
                class="prc-ActionList-Spacer-vgXtU"
              />
              <span
                class="prc-ActionList-ActionListSubContent-Zcjdz"
                data-component="ActionList.Item--DividerContainer"
              >
                <span
                  class="prc-ActionList-ItemLabel-hYyH7"
                  id=":r25:--label"
                >
                  Sub Item
                </span>
              </span>
            </a>
          </li>
        </ul>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`NavList.Item with NavList.SubNav > has active styles if SubNav contains the current item and is closed 1`] = `
<div>
  <nav>
    <ul
      class="prc-ActionList-ActionList--dkHj"
      data-dividers="false"
      data-variant="inset"
    >
      <li
        class="prc-ActionList-ActionListItem-jXerx"
        data-active="true"
        data-has-description="false"
        data-has-subitem="true"
      >
        <button
          aria-controls=":r1u:"
          aria-expanded="false"
          aria-labelledby=":r1t:--label :r1t:--trailing-visual "
          class="prc-ActionList-ActionListContent-Q8eQr"
          data-size="medium"
          id=":r1t:"
          style="--subitem-depth: 0;"
          tabindex="0"
          type="button"
        >
          <span
            class="prc-ActionList-Spacer-vgXtU"
          />
          <span
            class="prc-ActionList-ActionListSubContent-Zcjdz"
            data-component="ActionList.Item--DividerContainer"
          >
            <span
              class="prc-ActionList-ItemLabel-hYyH7"
              id=":r1t:--label"
            >
              Item
            </span>
            <span
              class="prc-ActionList-TrailingVisual-8LMjP prc-ActionList-VisualWrap-MGVnU"
              id=":r1t:--trailing-visual"
            >
              <svg
                aria-hidden="true"
                class="octicon octicon-chevron-down prc-ActionList-ExpandIcon-6JuzS"
                display="inline-block"
                fill="currentColor"
                focusable="false"
                height="16"
                overflow="visible"
                style="vertical-align: text-bottom;"
                viewBox="0 0 16 16"
                width="16"
              >
                <path
                  d="M12.78 5.22a.749.749 0 0 1 0 1.06l-4.25 4.25a.749.749 0 0 1-1.06 0L3.22 6.28a.749.749 0 1 1 1.06-1.06L8 8.939l3.72-3.719a.749.749 0 0 1 1.06 0Z"
                />
              </svg>
            </span>
          </span>
        </button>
        <ul
          aria-labelledby=":r1t:"
          class="prc-ActionList-SubGroup-4pVPh"
          id=":r1u:"
        >
          <li
            class="prc-ActionList-ActionListItem-jXerx"
            data-active="true"
            data-has-description="false"
          >
            <a
              aria-current="page"
              aria-labelledby=":r20:--label  "
              class="prc-ActionList-ActionListContent-Q8eQr prc-Link-Link-O33ci"
              data-size="medium"
              href="#"
              id=":r20:"
              style="--subitem-depth: 1;"
              tabindex="0"
            >
              <span
                class="prc-ActionList-Spacer-vgXtU"
              />
              <span
                class="prc-ActionList-ActionListSubContent-Zcjdz"
                data-component="ActionList.Item--DividerContainer"
              >
                <span
                  class="prc-ActionList-ItemLabel-hYyH7"
                  id=":r20:--label"
                >
                  Sub Item
                </span>
              </span>
            </a>
          </li>
        </ul>
      </li>
    </ul>
  </nav>
</div>
`;
